ml/nn/attention.go

```diff
@@ -0,0 +1,22 @@
+package nn
+
+import (
+	"github.com/ollama/ollama/ml"
+)
+
+func Attention(ctx ml.Context, query, key, value, mask ml.Tensor, scale float64) ml.Tensor {
+	if sdpa, ok := query.(ml.ScaledDotProductAttention); ok {
+		return sdpa.ScaledDotProductAttention(ctx, key, value, mask, scale)
+	} else {
+		kq := key.MulmatFullPrec(ctx, query)
```

Comment:
Should we be validating the tensor shapes here?

Reply:
Based on the information that we have here, I think the main thing that we could do is check that the corresponding dimensions of the input tensors match. For example, the sequence length of K matches V and the mask. Is that the kind of thing that you were thinking?

Reply:
That's pretty much what I was thinking, not a requirement by any means but maybe nice to have

Reply:
I think it's a good idea, it certainly will make debugging new models or backends easier compared to the eventual GGML error.